File: go/idkollen/ftn.go
package idkollen

import (
	"context"
	"encoding/json"
	"fmt"
	"time"
)

// --- FTN request types ---

type FtnAuthRequest struct {
	RedirectUrl    *string `json:"redirectUrl,omitempty"`
	RequestPhone   *bool   `json:"requestPhone,omitempty"`
	RequestEmail   *bool   `json:"requestEmail,omitempty"`
	RequestAddress *bool   `json:"requestAddress,omitempty"`
	RefId          *string `json:"refId,omitempty"`
}

// --- FTN status types ---

type FtnStatus interface {
	isFtnStatus()
}

type FtnPending struct {
	Id    string  `json:"id"`
	RefId *string `json:"refId"`
	Url   string  `json:"url"`
}

type FtnCompleted struct {
	Id        string  `json:"id"`
	RefId     *string `json:"refId"`
	Ssn       string  `json:"ssn"`
	Name      string  `json:"name"`
	GivenName string  `json:"givenName"`
	Surname   string  `json:"surname"`
	Phone     *string `json:"phone"`
	Email     *string `json:"email"`
	Address   *string `json:"address"`
	BirthDate *string `json:"birthDate"`
	Pid       *string `json:"pid"`
	BankId    *string `json:"bankId"`
}

type FtnFailed struct {
	Id    string  `json:"id"`
	RefId *string `json:"refId"`
	Error string  `json:"error"`
}

func (*FtnPending) isFtnStatus()   {}
func (*FtnCompleted) isFtnStatus() {}
func (*FtnFailed) isFtnStatus()    {}

type ftnStatusEnvelope struct {
	Status string `json:"status"`
}

func unmarshalFtnStatus(data []byte) (FtnStatus, error) {
	var env ftnStatusEnvelope
	if err := json.Unmarshal(data, &env); err != nil {
		return nil, err
	}
	switch env.Status {
	case "PENDING":
		var v FtnPending
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	case "COMPLETED":
		var v FtnCompleted
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	case "FAILED":
		var v FtnFailed
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	default:
		return nil, fmt.Errorf("unknown ftn status: %s", env.Status)
	}
}

// --- FTN endpoint ---

type FtnEndpoint struct {
	client *IdkollenClient
}

func (e *FtnEndpoint) Auth(ctx context.Context, req FtnAuthRequest) (FtnStatus, error) {
	var raw json.RawMessage
	if err := e.client.post(ctx, "/v3/ftn/auth", req, &raw); err != nil {
		return nil, err
	}
	return unmarshalFtnStatus(raw)
}

func (e *FtnEndpoint) AgeVerification(ctx context.Context, req AgeVerificationRequest) (AgeVerificationStatus, error) {
	var raw json.RawMessage
	if err := e.client.post(ctx, "/v3/ftn/age-verification", req, &raw); err != nil {
		return nil, err
	}
	return unmarshalAgeVerificationStatus(raw)
}

func (e *FtnEndpoint) AuthStatus(ctx context.Context, id string) (FtnStatus, error) {
	var raw json.RawMessage
	if err := e.client.get(ctx, "/v3/ftn/auth/"+id, &raw); err != nil {
		return nil, err
	}
	return unmarshalFtnStatus(raw)
}

func (e *FtnEndpoint) AgeVerificationStatus(ctx context.Context, id string) (AgeVerificationStatus, error) {
	var raw json.RawMessage
	if err := e.client.get(ctx, "/v3/ftn/age-verification/"+id, &raw); err != nil {
		return nil, err
	}
	return unmarshalAgeVerificationStatus(raw)
}

func (e *FtnEndpoint) CancelAuth(ctx context.Context, id string) error {
	return e.client.delete(ctx, "/v3/ftn/auth/"+id)
}

func (e *FtnEndpoint) CancelAgeVerification(ctx context.Context, id string) error {
	return e.client.delete(ctx, "/v3/ftn/age-verification/"+id)
}

func (e *FtnEndpoint) WaitForAuth(ctx context.Context, id string, opts PollOptions) (FtnStatus, error) {
	interval := opts.Interval
	if interval == 0 {
		interval = 2 * time.Second
	}
	for {
		status, err := e.AuthStatus(ctx, id)
		if err != nil {
			return nil, &WaitError{Err: err}
		}
		if _, ok := status.(*FtnPending); !ok {
			return status, nil
		}
		select {
		case <-ctx.Done():
			return nil, &WaitError{Timeout: true}
		case <-time.After(interval):
		}
	}
}

func (e *FtnEndpoint) WaitForAgeVerification(ctx context.Context, id string, opts PollOptions) (AgeVerificationStatus, error) {
	interval := opts.Interval
	if interval == 0 {
		interval = 2 * time.Second
	}
	for {
		status, err := e.AgeVerificationStatus(ctx, id)
		if err != nil {
			return nil, &WaitError{Err: err}
		}
		if _, ok := status.(*AgeVerificationPending); !ok {
			return status, nil
		}
		select {
		case <-ctx.Done():
			return nil, &WaitError{Timeout: true}
		case <-time.After(interval):
		}
	}
}

File: ruby/lib/idkollen/ftn.rb
# frozen_string_literal: true

module Idkollen
  module Ftn
    # --- Request ---

    class AuthRequest
      attr_reader :redirect_url, :request_phone, :request_email, :request_address, :ref_id

      def initialize(redirect_url: nil, request_phone: nil, request_email: nil,
                     request_address: nil, ref_id: nil)
        @redirect_url    = redirect_url
        @request_phone   = request_phone
        @request_email   = request_email
        @request_address = request_address
        @ref_id          = ref_id
      end

      def to_h
        {
          redirectUrl:    @redirect_url,
          requestPhone:   @request_phone,
          requestEmail:   @request_email,
          requestAddress: @request_address,
          refId:          @ref_id,
        }.compact
      end
    end

    # --- Status variants ---

    class Pending
      attr_reader :id, :ref_id, :url

      def initialize(id:, url:, ref_id: nil)
        @id = id; @ref_id = ref_id; @url = url
      end

      def deconstruct_keys(_) = { id: @id, ref_id: @ref_id, url: @url }

      def self.from_json(d) = new(id: d['id'], ref_id: d['refId'], url: d['url'])
    end

    class Completed
      attr_reader :id, :ref_id, :ssn, :name, :given_name, :surname,
                  :phone, :email, :address, :birth_date, :pid, :bank_id

      def initialize(id:, ssn:, name:, given_name:, surname:,
                     ref_id: nil, phone: nil, email: nil, address: nil,
                     birth_date: nil, pid: nil, bank_id: nil)
        @id = id; @ref_id = ref_id; @ssn = ssn; @name = name
        @given_name = given_name; @surname = surname
        @phone = phone; @email = email; @address = address
        @birth_date = birth_date; @pid = pid; @bank_id = bank_id
      end

      def deconstruct_keys(_)
        { id: @id, ref_id: @ref_id, ssn: @ssn, name: @name, given_name: @given_name,
          surname: @surname, phone: @phone, email: @email, address: @address,
          birth_date: @birth_date, pid: @pid, bank_id: @bank_id }
      end

      def self.from_json(d)
        new(id: d['id'], ref_id: d['refId'], ssn: d['ssn'], name: d['name'],
            given_name: d['givenName'], surname: d['surname'],
            phone: d['phone'], email: d['email'], address: d['address'],
            birth_date: d['birthDate'], pid: d['pid'], bank_id: d['bankId'])
      end
    end

    class Failed
      attr_reader :id, :ref_id, :error

      def initialize(id:, error:, ref_id: nil)
        @id = id; @ref_id = ref_id; @error = error
      end

      def deconstruct_keys(_) = { id: @id, ref_id: @ref_id, error: @error }

      def self.from_json(d) = new(id: d['id'], ref_id: d['refId'], error: d['error'])
    end

    module Status
      def self.from_json(d)
        case d['status']
        when 'PENDING'   then Pending.from_json(d)
        when 'COMPLETED' then Completed.from_json(d)
        when 'FAILED'    then Failed.from_json(d)
        else raise ArgumentError, "Unknown ftn status: #{d['status']}"
        end
      end
    end

    # --- Endpoint ---

    class Endpoint
      def initialize(transport)
        @transport = transport
      end

      def auth(req)
        Status.from_json(@transport.post('/v3/ftn/auth', req.to_h))
      end

      def age_verification(req)
        Idkollen::AgeVerification::Status.from_json(@transport.post('/v3/ftn/age-verification', req.to_h))
      end

      def auth_status(id) = Status.from_json(@transport.get("/v3/ftn/auth/#{id}"))

      def age_verification_status(id)
        Idkollen::AgeVerification::Status.from_json(@transport.get("/v3/ftn/age-verification/#{id}"))
      end

      def cancel_auth(id)             = @transport.delete("/v3/ftn/auth/#{id}")
      def cancel_age_verification(id) = @transport.delete("/v3/ftn/age-verification/#{id}")

      def wait_for_auth(id, opts: PollOptions.new)
        poll_ftn(opts) { auth_status(id) }
      end

      def wait_for_age_verification(id, opts: PollOptions.new)
        poll_age(opts) { age_verification_status(id) }
      end

      private

      def poll_ftn(opts)
        deadline = Time.now + opts.timeout
        loop do
          status = yield
          return status unless status.is_a?(Pending)
          raise WaitError.new(timeout: true) if Time.now >= deadline
          sleep opts.interval
        end
      end

      def poll_age(opts)
        deadline = Time.now + opts.timeout
        loop do
          status = yield
          return status unless status.is_a?(Idkollen::AgeVerification::Pending)
          raise WaitError.new(timeout: true) if Time.now >= deadline
          sleep opts.interval
        end
      end
    end
  end
end
